Late arriving data (#4386)
influxdata/docs-v2: content/resources/how-to-guides/late-arriving-data.md

* An invokable script:
    * water_level_process.flux

Also imagine the following scenario. You’re writing a fake dataset. This fake dataset contains water levels at 100 different locations. You will be computing the mean water level at each location over 1 minute windows. You can handle data arriving up to 1 hour late. Data for each location is being written every 10s. Additionally every 10s a late data point is written somewhere in the last 1 hour for each location. 

## Overview 

Before diving into the code, take a high level look at the logic of the Flux scripts.

![late arriving data architecture](images/late-arriving-data.png "image_tooltip")

The water_level_checksum.flux is a task. It runs every minute. It counts the number of points that exist in the water_level_raw bucket (new count) and compares that count against the count in the water_level_checksum bucket (old count). If the new count from the water_level_raw bucket isn’t equal to the count from the water_level_checksum bucket then an invokable script is called. Specifically, the water_level_process.flux script is called. This script is responsible for recalculating the old count and aggregation.   

> ## Flux scripts in detail: water_level_process.flux
> 
> The water_level_process.flux script is an invokable script. It’s responsible for two things:

Reviewer: ```suggestion
## Flux scripts in detail

- [water_level_process.flux](#water_level_processflux)
- [water_level_checksum.flux](#water_level_checksumflux)

### water_level_process.flux

`water_level_process.flux` is an invokable script that does two things:
```

> 1. Computing the mean value over a parameterized start and stop times. This mean value is written to the water_level_mean bucket. 
> 2. Calculating the count or total numper of points over parameterized start and stop times. This count is written to the water_level_checksum bucket. 

Reviewer: ```suggestion
1. Computes the mean of values in the time range defined by the `start` and `stop` script parameters and writes the computed mean to the **water_level_mean** bucket. 
2. Computes the count or total number of points in the time range defined by the `start` and `stop` script parameters and writes the count to the **water_level_checksum** bucket. 
```

The water_leve_process.flux script contains the following code: \
`// Compute the mean for the window`

```
from(bucket: "water_level_raw")
    |> range(start: params.start, stop: params.stop)
    |> mean()
    |> to(bucket: "water_level_mean", timeColumn: "_stop")
    |> yield(name: "means")

// Compute and store new checksum for this window
from(bucket: "water_level_raw")
    |> range(start: params.start, stop: params.stop)
    |> group(columns: ["_measurement", "_field", "_stop"])
    |> count()
    |> to(bucket: "water_level_checksum", timeColumn: "_stop")
    |> yield(name: "checksums")
```

> 

Reviewer: ~~~suggestion
```js
// Compute and store the mean for the window
from(bucket: "water_level_raw")
    |> range(start: params.start, stop: params.stop)
    |> mean()
    |> to(bucket: "water_level_mean", timeColumn: "_stop")
    |> yield(name: "means")

// Compute and store the new checksum for this window
from(bucket: "water_level_raw")
    |> range(start: params.start, stop: params.stop)
    |> group(columns: ["_measurement", "_field", "_stop"])
    |> count()
    |> to(bucket: "water_level_checksum", timeColumn: "_stop")
    |> yield(name: "checksums")
```
~~~

> ## Flux scripts in detail: water_level_checsum.flux
> 
> The water_level_process.flux script is a task. This task is responsible for:

Reviewer: ```suggestion
## water_level_checsum.flux

`water_level_process.flux` is a task that does the following:
```

> 1. Counting the number of points in the water_level_raw bucket (new count) for the last hour across 1minute windows.
> 2. Invoking the water_level_process.flux invokable script to calculate a new mean and a new count across 1 minute windows. 
> 3. Gathering the previous count in the water_level_checksum bucket (old count) for the last hour. 
> 4. Comparing an old count against a new count with a join.  
> 5. Filtering for where the counts do not match.
> 6. Invoking the water_level_process.flux invokable script to recompute the mean and count for every 1 min window where the counts don't match. 

Reviewer: ```suggestion
1. Counts the number of points in the **water_level_raw** bucket (new count) for the last hour across one minute windows.
2. Invokes the `water_level_process.flux` invokable script to calculate a new mean and a new count across one minute windows. 
3. Gathers the previous count in the **water_level_checksum** bucket (old count) for the last hour. 
4. Joins the old and new streams and compares the old count against a new count.  
5. Filters for counts that do not match.
6. Invokes the `water_level_process.flux` invokable script to recompute the mean and count for every one minute window where the counts do not match. 
```

> The task runs at 1m intervals specified by the `every` parameter:
> 
> 
> ```
> option task = {name: "water_level_checksum", every: 1m, offset: 10s}
> ```
> 
> 
> This `every` parameter also determines the window period for which the mean and counts will be computed over. 
> 
> The invokeScripts() function is a custom function that invokes the water_level_process.flux invokable script. \
> `invokeScript = (start, stop) =>`
> 
> 

Reviewer: ```suggestion
#### Task details

- The `task` option provides configuration settings for the task:
    - `name`: Provides a name for the task.
    - `every`: Defines how often the task runs (every one minute) and, in this case, the window interval used to compute means and counts.
    - `offset`: Defines how much time to wait before executing the task. _The offset does not change the time range queried by the task. _
- `invokeScripts()` is a custom function that invokes the `water_level_process.flux` invokable script.
    - `start` and `stop` parameters are required.
    - `scriptID` is required.
    - Store your InfluxDB API token as an InfluxDB secret and use the `secrets` package to retrieve the token.
```

```
    requests.post(
        // The script ID here has been hardcoded here
        url: "https://eastus-1.azure.cloud2.influxdata.com/api/v2/scripts/095fabd404108000/invoke",

        headers: ["Authorization": "Token ${token}", "Accept": "application/json", "Content-Type": "application/json"],
        body: json.encode(v: {params: {start: string(v: start), stop: string(v: stop)}}),
    )

> ```

Reviewer: I'd parameterize the script ID.

~~~suggestion
```
option task = {name: "water_level_checksum", every: 1m, offset: 10s}

invokeScript = (start, stop, scriptID) =>
    requests.post(
        url: "https://cloud2.influxdata.com/api/v2/scripts/${scriptID}/invoke",

        headers: ["Authorization": "Token ${token}", "Accept": "application/json", "Content-Type": "application/json"],
        body: json.encode(v: {params: {start: string(v: start), stop: string(v: stop)}}),
    )
```
~~~

We also need to instruct them where to find the script ID.

> Remember this invokable script is responsible for both calculating the mean and count and writing it to the water_level_mean and water_level_checksum buckets respectively.  You must pass in values for the start and stop parameters. The token is imported with the secrets manager. 

Reviewer: ```suggestion
```

Next, query the water_level_checksum bucket for the last hour: 

```
oldCounts =
    from(bucket: "water_level_checksum")
        |> range(start: start, stop: stop)
        |> group(columns: ["_measurement", "_field"])
```

Remember the start and stop times here are equal to - and now() truncated to the minute. 

Now join the old counts and new counts together. You also filter for when the counts differ. If they do differ, then there will be records in the response that can be mapped over. Map over those records to recalculate the mean and count by invoking the level_water_process.flux script: 

```
experimental.join(
    left: oldCounts,
    right: newCounts,
    fn: (left, right) => ({left with old_count: left._value, new_count: right._value}),
)
    // Recompute any windows where the checksum is different
    |> filter(fn: (r) => r.old_count != r.new_count)
    |> map(
        fn: (r) => {
            response = invokeScript(start: date.sub(d: every, from: r._time), stop: r._time)

            return {r with code: response.statusCode}
        },
    )
    |> yield(name: "diffs")

> ```

Reviewer: Explain the task in the proposed "Task details" section, but show the task in it's entirety.

Author: Thank you! All comments have been addressed.
